2-collectors/scala-stream-collector/core/src/main/scala/com.snowplowanalytics.snowplow.collectors.scalastream/CollectorService.scala

```diff
@@ -343,6 +342,74 @@ class CollectorService(
     if (pixelExpected) List(`Cache-Control`(`no-cache`, `no-store`, `must-revalidate`))
     else Nil
 
+  /**
+   * Determines the cookie domain to be used by inspecting the Origin header of the request
+   * and trying to find a match in the list of domains specified in the config file.
+   * The Origin header may include multiple domains. The first matching domain is used.
+   * If no match is found, the cookie domain is not set.
+   */
+  def cookieDomain(request: HttpRequest, domains: Option[List[String]], fallbackDomain: Option[String]): Option[String] = {
+    domains match {
+      case Some(domainList) =>
+        request.headers.find {
+          case `Origin`(_) => true
+          case _ => false
+        } match {
+          case Some(`Origin`(origins)) =>
+            val originHosts = extractHosts(origins)
+            val domainRegex = domainList.map(toRegex)
+            val matchingHosts = for {
+              domain <- domainRegex
+              matchingHosts <- originHosts.find {
+                case x if x.matches(domain) => true
+                case _ => false
+              }
+            } yield matchingHosts
+            matchingHosts.headOption match {
+              case Some(matchingValue) =>
+                val domainToUse = domainRegex.find {
+                  case x if matchingValue.matches(x) => true
+                  case _ => false
+                }
+                domainToUse.flatMap { d => Some(toDomain(d)) }
+              case None => fallbackDomain
+            }
+          case _ => fallbackDomain
+        }
+      case None => fallbackDomain
+    }
+  }
+
+  /** Extracts the host names from a list of values in the request's Origin header. */
+  def extractHosts(origins: Seq[HttpOrigin]): Seq[String] = {
+    origins.map(origin => origin.host.host.address())
```

Comment:
doesn't `origin.host.value` give us what we want?

Reply:
:art: no need for curlies also

Reply:
> doesn't `origin.host.value` give us what we want?

No, it returns the full host, including the port.